java/core/src/test/org/apache/orc/impl/TestSchemaEvolution.java

```diff
@@ -2265,4 +2265,72 @@ public void testEvolutionToTimestamp() throws Exception {
       TimeZone.setDefault(oldDefault);
     }
   }
+
+  @Test
+  public void doubleToTimeStampPositiveOverflow() throws Exception {
+    floatAndDoubleToTimeStampOverflow(340282347000000000000000000000000000000000.0, "double");
+  }
+
+  @Test
+  public void floatToTimeStampPositiveOverflow() throws Exception {
+    floatAndDoubleToTimeStampOverflow(340282347000000000000000000000000000000000.0, "float");
+  }
+
+  @Test
+  public void doubleToTimeStampNegativeOverflow() throws Exception {
+    floatAndDoubleToTimeStampOverflow(-340282347000000000000000000000000000000000.0, "double");
+  }
+
+  @Test
+  public void floatToTimeStampNegativeOverflow() throws Exception {
+    floatAndDoubleToTimeStampOverflow(-340282347000000000000000000000000000000000.0, "float");
+  }
+
+  private void floatAndDoubleToTimeStampOverflow(double value, String typeInFileSchema)
+      throws Exception {
+    final ZoneId READER_ZONE = ZoneId.of("Australia/Sydney");
+
+    TypeDescription fileSchema =
+        TypeDescription.fromString(String.format("struct<c1:%s>", typeInFileSchema));
+    Writer writer = OrcFile.createWriter(testFilePath,
+        OrcFile.writerOptions(conf).setSchema(fileSchema).stripeSize(10000));
+
+    VectorizedRowBatch batch = fileSchema.createRowBatchV2();
+    DoubleColumnVector fl1 = (DoubleColumnVector) batch.cols[0];
+
+    int batchSize = batch.getMaxSize();
+
+    for (int r = 0; r < 1024; ++r) {
+      int row = batch.size++;
+      fl1.vector[row] = value;
+
+      if (batch.size == batchSize) {
+        writer.addRowBatch(batch);
+        batch.reset();
+      }
+    }
+    if (batch.size != 0) {
+      writer.addRowBatch(batch);
+    }
+    writer.close();
+
+    final TimeZone oldDefault = TimeZone.getDefault();
+    TimeZone.setDefault(TimeZone.getTimeZone(READER_ZONE));
+
+    TypeDescription readerSchema = TypeDescription.fromString("struct<c1:timestamp>");
+    VectorizedRowBatch batchTimeStamp = readerSchema.createRowBatchV2();
+    TimestampColumnVector t1 = (TimestampColumnVector) batchTimeStamp.cols[0];
+
+    OrcFile.ReaderOptions options = OrcFile.readerOptions(conf);
+    Reader.Options rowOptions = new Reader.Options().schema(readerSchema);
+
+    try (Reader reader = OrcFile.createReader(testFilePath, options);
+      RecordReader rows = reader.rows(rowOptions)) {
+      assertTrue(rows.nextBatch(batchTimeStamp));
+
+      assertTrue(String.format("nanos should be > 0, instead it's: %d", t1.nanos[0]),
```

Comment:
This test case is much too loose. Let me take a pass at strengthening it.

Reply:
yeah, test case is cleaner and covers+checks lots of values and use cases now, +1